File: src/main/resources/static/js/common/util/pageUtil.js
/*
    페이징 공식

    page		// 현재 페이지
    size		// 페이지 크기 (start~end)
    limit		// 페이지당 보여줄 아이템 수
    totalCount	// 총 아이템 수

    totalPages	// 총 페이지 (totalCount -1 / limit) + 1
    startIndex	// 시작번호 (page - 1) * limit
    endIndex	// 끝번호 Math.min(startIndex + limit - 1, totalCount - 1)

    프론트 -> 서버로 보내야 하는 값 : (page, size, limit)
*/



const getPageParam = (page) => {

    return {
        page: page,
        size: 5,
        limit: 10,
    }

}

// 페이징 클릭 시 콜백 함수를 searchPage로 선언 해야 한다
const handlePageClick = (el) => {
    const page = parseInt(el.dataset.page);
    searchPage(page);
}

const prev = (el) => {
    let page = 1;
    const startPage = parseInt(el.dataset.startPage);
    const size = parseInt(el.dataset.size);

    page = Math.max(startPage - size, 1);

    searchPage(page);
}

const next = (el) => {
    let page = 1;
    const totalPages = parseInt(el.dataset.totalPages);
    const endPage = parseInt(el.dataset.endPage);

    page = Math.min(endPage + 1, totalPages);

    searchPage(page);
}

File: src/main/java/com/luckyvicky/common/util/PageUtil.java
package com.luckyvicky.common.util;

import com.luckyvicky.common.vo.PageVO;
import com.luckyvicky.web.client.common.dto.BaseSearchDTO;
import org.springframework.stereotype.Component;

@Component
public class PageUtil {


    // 파라미터 다형성
    public PageVO getPageVO(BaseSearchDTO searchDTO, Long totalCount) {

        // 페이징 값 초기화
        int page = searchDTO.getPage();
        int size = searchDTO.getSize();
        int limit = searchDTO.getLimit();

        int totalPages = (int) (((totalCount - 1) / limit) + 1);
        int startIndex = (page - 1) * limit; // offset
        int endIndex = (int) Math.min(startIndex + limit - 1, totalCount - 1);

        // 페이지 버튼 (프론트용)
        int startPage = ((page - 1) / size) * size + 1;
        int endPage = Math.min(startPage + size - 1, totalPages);



        return PageVO.builder()
                .page(page)
                .size(size)
                .limit(limit)
                .totalCount(totalCount)
                .totalPages(totalPages)
                .startIndex(startIndex)
                .endIndex(endIndex)
                .startPage(startPage)
                .endPage(endPage)
                .build();
    }

}
